pkg/certificateprovider/azurekeyvault/provider.go

```diff
@@ -0,0 +1,368 @@
+)
+
+type keyvaultObject struct {
+	content string
+	version string
+}
+
+func GetCertificatesContent(ctx context.Context, attrib map[string]string) ([]types.CertificateFile, error) {
+
+	keyvaultName := types.GetKeyVaultName(attrib)
+	cloudName := types.GetCloudName(attrib)
+	tenantID := types.GetTenantID(attrib)
+	workloadIdentityClientID := types.GetClientID(attrib)
+
+	if keyvaultName == "" {
+		return nil, fmt.Errorf("keyvaultName is not set")
+	}
+	if tenantID == "" {
+		return nil, fmt.Errorf("tenantID is not set")
+	}
+	if workloadIdentityClientID == "" {
+		return nil, fmt.Errorf("clientId is not set")
+	}
+
+	azureCloudEnv, err := ParseAzureEnvironment(cloudName)
+	if err != nil {
+		return nil, fmt.Errorf("cloudName %s is not valid, error: %w", cloudName, err)
+	}
+
+	// 1. cleaning up keyvault objects definition
+	objectsStrings := types.GetCertificates(attrib)
+	if objectsStrings == "" {
+		return nil, fmt.Errorf("objects is not set")
+	}
+	logrus.Infof("objects string defined in secret provider class, objects %v", objectsStrings)
+
+	objects, err := types.GetCertificatesArray(objectsStrings)
+	if err != nil {
+		return nil, fmt.Errorf("failed to yaml unmarshal objects, error: %w", err)
+	}
+	logrus.Infof("unmarshaled objects yaml, objectsArray %v", objects.Array)
+
+	keyVaultCerts := []types.KeyVaultCertificate{}
+	for i, object := range objects.Array {
+		var keyVaultCert types.KeyVaultCertificate
+		err = yaml.Unmarshal([]byte(object), &keyVaultCert)
+		if err != nil {
+			return nil, fmt.Errorf("unmarshal failed for keyVaultCerts at index %d, error: %w", i, err)
+		}
+		// remove whitespace from all fields in keyVaultCert
+		formatKeyVaultCertificate(&keyVaultCert)
+		if err = validate(keyVaultCert); err != nil {
+			return nil, wrapObjectTypeError(err, keyVaultCert.CertificateName, keyVaultCert.CertificateVersion)
+		}
+
+		keyVaultCerts = append(keyVaultCerts, keyVaultCert)
+	}
+
+	logrus.Infof("unmarshaled key vault objects, keyVaultObjects %v , count %v", keyVaultCerts, len(keyVaultCerts))
+
+	if len(keyVaultCerts) == 0 {
+		return nil, nil
+	}
+
+	// 2. initialize keyvault client
+
+	vaultURL, err := getVaultURL(keyvaultName, azureCloudEnv.KeyVaultDNSSuffix)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get vault")
+	}
+	logrus.Infof("vaultName %v, vaultURL %v", keyvaultName, *vaultURL)
+
+	kvClient, err := initializeKvClient(ctx, azureCloudEnv.KeyVaultEndpoint, azureCloudEnv.ActiveDirectoryEndpoint, tenantID, workloadIdentityClientID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get keyvault client")
+	}
+
+	// 3. for each object , get content bytes
+	files := []types.CertificateFile{}
+	for _, keyVaultCert := range keyVaultCerts {
+		logrus.Infof("fetching object from key vault, certName %v,  keyvault %v", keyVaultCert.CertificateName, keyvaultName)
+
+		resolvedKvCerts, err := resolveCertificateVersions(ctx, kvClient, keyVaultCert, *vaultURL)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
Is `resolveCertificateVersions` required for ratify? That exists in CSI provider because we support getting `n` versions of an object from keyvault. I don't think that's relevant here.

Reply:
removed. thanks